### PR TITLE
[VECTOR_FLOAT16] Implement feature extension and version negotiation for Vector v2 support

### PR DESCRIPTION
**Pull request by muskan124947**
## Description

For upcoming VECTOR_FLOAT16 support in the Microsoft JDBC Driver for SQL Server, this PR introduces the initial framework for vector type support handling on the client side, enabling controlled negotiation of vector capabilities between client and server.

The changes focus on making vector support explicit, extensible, and future-proof, in preparation for additional vector versions and data types.

## Vector Version Negotiation Flow

- The client vector setting is parsed and validated:
```
off – vector support disabled
v1 – float32 vector support
v2 – float32 and float16 vector support
```
- If either the client disables vectors or the server does not support vectors, vector support is disabled.
- The client and server determine their maximum supported vector versions.
- When both support vectors, the negotiated version is the minimum of the client and server versions.
- The negotiated version is used for all vector-related operations for the connection.

## Impact

- Backward compatibility: Clients using only float32 continue to work without changes.
- Performance: Float16 support reduces memory usage and network payload for supported operations.
- Extensibility: Framework allows easy addition of future vector types and versions.

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 1]
```diff
@@ -7152,6 +7193,57 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
...
+     * - If either client or server is "off", negotiated version is "off"
+     * - If both support v2, negotiated version is v2
+     * - If both support v1, negotiated version is v1
+     * - Otherwise, use the minimum supported version
+     *
+     * @param clientVectorSupport The client's vector type support setting
+     * @param serverVersion       The server's supported vector version
+     * @return The negotiated vector version
+     */
+    private byte negotiateVectorVersion(String clientVectorSupport, byte serverVersion) {
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: We should check and validate connection string value for vectorSupport option as a string only once and then map it to an enum value and use it everywhere instead of string values.

When we send the vector FE to server as part of pre-login, we should compare the connection string value of vectorSupport and driver's current vector version, do the min of it and send that to server. This simplifies trhe negoation.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerConnection.java [hunk 2]
```diff
@@ -658,7 +658,10 @@ CallableStatement prepareCall(String sql, int nType, int nConcur, int nHold,
      *
      * @param vectorTypeSupport
      * A string that indicates the vector type support during connection initialization.
-     * Valid values are "off" (vector types are returned as strings) and "v1" (vectors of type FLOAT32 are returned as vectors). 
+     * Valid values are :
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerConnection.java:661]: Could we map the user supplied value from connection string to an enum type that has each vector type support byte defined. Once we map this then we can use that byte value directly to send to server without needing additional switch cases.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 3]
```diff
@@ -5899,14 +5908,28 @@ int writeUserAgentFeatureRequest(boolean write, /* if false just calculates the
...
     int writeVectorSupportFeatureRequest(boolean write,
             TDSWriter tdsWriter) throws SQLServerException {
-        if (VECTOR_SUPPORT_OFF.equalsIgnoreCase(vectorTypeSupport)) {
+
+        // Initialize vectorTypeSupportEnum if not already done.
+        if (vectorTypeSupportEnum == null) {
+            try {
+                vectorTypeSupportEnum = VectorTypeSupport.valueOfString(vectorTypeSupport);
+            } catch (SQLServerException e) {
+                // Fallback to value OFF if invalid value is provided.
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: we should just throw the error back to user instead of silently falling back

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 4]
```diff
@@ -7118,7 +7141,20 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
                 if (0 == serverSupportedVectorVersion || serverSupportedVectorVersion > TDS.MAX_VECTORSUPPORT_VERSION) {
                     throw new SQLServerException(SQLServerException.getErrString("R_InvalidVectorVersionNumber"), null);
                 }
-                serverSupportsVector = true;
+                // Negotiate the vector version between client and server
+                negotiatedVectorVersion = negotiateVectorVersion(vectorTypeSupportEnum, serverSupportedVectorVersion);
```
Comment (machavan) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: The server would have already done the negotiation, do we really need to do negotiation again or is it just a  conservative/safety code ?

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/VectorTest.java [hunk 5]
```diff
@@ -1733,21 +1733,280 @@ void validatePreparedStatementMetaData() throws SQLException {
...
+         * - Case variations (uppercase, mixed case)
+         * - Default behavior when not specified
+         */
+        @Test
+        @DisplayName("Test valid vectorTypeSupport edge cases")
+        public void testValidVectorTypeSupportEdgeCases() throws SQLException {
+            // Test cases: [connectionStringValue, description, expectedNormalizedValue]
+            // The driver trims whitespace and is case-insensitive
+            String[][] validCases = {
+                {" v1 ", "whitespace-padded v1", "v1"},
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/VectorTest.java:1915]: trailing spaces ?

**Review by machavan — DISMISSED** [review]

**Review by divang — COMMENTED** [review]
Do we need a way to identify different statement types: Statement, PreparedStatement, and CallableStatement , so that users can analyze and compare the performance differences among them?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 6]
```diff
@@ -7154,6 +7182,45 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
...
+     * - If either client or server is "off", negotiated version is "off"
+     * - If both support v2, negotiated version is v2
+     * - If both support v1, negotiated version is v1
+     * - Otherwise, use the minimum supported version
+     *
+     * @param clientVectorSupportEnum The client's vector type support setting
+     * @param serverVersion           The server's supported vector version
+     * @return The negotiated vector version
+     */
+    private byte negotiateVectorVersion(VectorTypeSupport clientVectorSupportEnum, byte serverVersion) {
```
Comment (divang) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Is this method in use negotiateVectorVersion() ?

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/VectorTest.java [hunk 7]
```diff
@@ -1733,21 +1733,281 @@ void validatePreparedStatementMetaData() throws SQLException {
...
+
+            // Load VectorTypeSupport enum via reflection (it's package-private)
+            vectorTypeSupportClass = Class.forName("com.microsoft.sqlserver.jdbc.VectorTypeSupport");
+            VECTOR_TYPE_OFF = Enum.valueOf((Class<Enum>) vectorTypeSupportClass, "OFF");
+            VECTOR_TYPE_V1 = Enum.valueOf((Class<Enum>) vectorTypeSupportClass, "V1");
+            VECTOR_TYPE_V2 = Enum.valueOf((Class<Enum>) vectorTypeSupportClass, "V2");
+        }
+
+        // ============================================================================
+        // Direct Unit Tests for negotiateVectorVersion method
```
Comment (divang) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/VectorTest.java]: If negotiateVectorVersion() method not is use then we can remove the test cases for that.

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/VectorTest.java [hunk 8]
```diff
@@ -773,8 +773,8 @@ void testInvalidVectorTypeSupportConnectionProperty() throws SQLException {
                 fail("Expected IllegalArgumentException for invalid vectorTypeSupport value, but none was thrown.");
             } catch (IllegalArgumentException e) {
                 exceptionThrown = true;
-                assertTrue(
-                        e.getMessage().contains("Incorrect connection string property for vectorTypeSupport: 1"),
+                String expectedMessage = "Invalid value for vectorTypeSupport: 1. Valid values are \"off\", \"v1\", or \"v2\".";
```
Comment (divang) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/VectorTest.java:776]: Do we need to match with R_invalidVectorTypeSupport constant?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 9]
```diff
@@ -5899,14 +5908,23 @@ int writeUserAgentFeatureRequest(boolean write, /* if false just calculates the
      */
     int writeVectorSupportFeatureRequest(boolean write,
             TDSWriter tdsWriter) throws SQLServerException {
-        if (VECTOR_SUPPORT_OFF.equalsIgnoreCase(vectorTypeSupport)) {
+
+        // Initialize vectorTypeSupportEnum if not already done.
+        if (vectorTypeSupportEnum == null) {
+            vectorTypeSupportEnum = VectorTypeSupport.valueOfString(vectorTypeSupport);
+        }
```
Comment (divang) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:5915]: why do need null check? if it is already initialized by default. 
private VectorTypeSupport vectorTypeSupportEnum = VectorTypeSupport.V1

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 10]
```diff
@@ -1136,15 +1146,14 @@ public void setVectorTypeSupport(String vectorTypeSupport) {
...
-        switch (vectorTypeSupport.trim().toLowerCase()) {
-            case VECTOR_SUPPORT_OFF:
-            case VECTOR_SUPPORT_V1:
-                this.vectorTypeSupport = vectorTypeSupport.toLowerCase();
-                break;
-            default:
-                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_invalidVectorTypeSupport"));
-                Object[] msgArgs = { vectorTypeSupport };
-                throw new IllegalArgumentException(form.format(msgArgs));
+        String normalizedValue = vectorTypeSupport.trim().toLowerCase(Locale.US);
```
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:1149]: Consider using Locale.ROOT instead of Locale.US here, since this is protocol-level normalization and should be culture-neutral and language-independent.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 11]
```diff
@@ -7114,11 +7132,21 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
                     throw new SQLServerException(SQLServerException.getErrString("R_unknownVectorSupportValue"), null);
                 }
 
+                // The server's FEATUREEXTACK response already contains the negotiated version
+                // (i.e. min(clientRequested, serverMax)), so we accept data[0] directly
+                // as the negotiated version — no client-side re-negotiation needed.
                 serverSupportedVectorVersion = data[0];
-                if (0 == serverSupportedVectorVersion || serverSupportedVectorVersion > TDS.MAX_VECTORSUPPORT_VERSION) {
+                negotiatedVectorVersion = data[0];
+                if (0 == negotiatedVectorVersion || negotiatedVectorVersion > TDS.MAX_VECTORSUPPORT_VERSION) {
```
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:7141]: getNegotiatedVectorVersion() documents 0 = off, but the ACK logic treats 0 as invalid. Please align the validation with the comment, either allow 0 or update the comment.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: If data.length == 0, accessing data[0] will throw ArrayIndexOutOfBoundsException before the intended validation. Please validate the length first, then log data[0] (lines 7131–7133).

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java (hunk 6 above)
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: negotiateVectorVersion() is currently unused. Are we planning to integrate it into the ACK handling logic?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java [hunk 12]
```diff
@@ -501,6 +501,64 @@ static DatetimeType valueOfString(String value) throws SQLServerException {
...
+        switch (value.trim().toLowerCase(Locale.US)) {
+            case "off":
+                return VectorTypeSupport.OFF;
+            case "v1":
+                return VectorTypeSupport.V1;
+            case "v2":
+                return VectorTypeSupport.V2;
+            default:
+                throwInvalid(value);
+                return VectorTypeSupport.OFF; // Never reached
```
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java]: remove the unreachable return VectorTypeSupport.OFF after throwInvalid()

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java [hunk 13]
```diff
@@ -501,6 +501,64 @@ static DatetimeType valueOfString(String value) throws SQLServerException {
...
+    public String getType() {
+        return type;
+    }
+
+    static VectorTypeSupport valueOfString(String value) throws SQLServerException {
+        if (value == null) {
+            throwInvalid(value);
+        }
+
+        switch (value.trim().toLowerCase(Locale.US)) {
```
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java:541]: Consider using Locale.ROOT instead of Locale.US here.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 14]
```diff
@@ -7154,6 +7182,45 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
...
+
+        // Negotiate using the minimum supported version
+        return (byte) Math.min(clientMaxVersion, serverVersion);
+
+    }
+
+    /**
+     * Returns the negotiated vector version between client and server.
+     *
+     * @return The negotiated vector version (0 = off, 1 = v1, 2 = v2)
```
Comment (Ananya2) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: getNegotiatedVectorVersion() documents 0 = off, but the ACK logic treats 0 as invalid. Please align the validation with the comment, either allow 0 or update the comment.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java (hunk 12 above)
Comment (muskan124947) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java]: Moved the unreachable [return VectorTypeSupport.OFF] out of the default case to after the switch, with a comment explaining it's only there to satisfy the compiler since [throwInvalid()] always throws
